### PR TITLE
[no-ticket] cached parent package annotation scanner

### PR DESCRIPTION
**Pull request by cingaldi**

### CODE REVIEW

File: src/main/java/com/glovoapp/ownership/scanning/CachedParentPackageAnnotationScanner.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    @Override
+    public Optional<A> scan(AnnotatedElement element) {
+        Optional<String> optionalPackageName =  PackageScanningUtils.getPackageName(element);
+
+        if (optionalPackageName.isPresent()) {
+            return cache.computeIfAbsent(optionalPackageName.get(), (k) -> delegate.scan(element));
+        }
+
+        return Optional.empty();
+
```
Comment (Jezorko) [inline src/main/java/com/glovoapp/ownership/scanning/CachedParentPackageAnnotationScanner.java:26]: Could be simplified:

```suggestion
        return PackageScanningUtils.getPackageName(element)
                                   .flatMap(packageName -> cache.computeIfAbsent(packageName,
                                       ignored -> delegate.scan(element)
                                   ));
```

File: src/main/java/com/glovoapp/ownership/OwnershipAnnotationDefinition.java [hunk 2]
```diff
@@ -53,7 +56,11 @@
     static <A extends Annotation> OwnershipAnnotationDefinition define(@NonNull final Class<A> annotationClass,
                                                                        @NonNull final Function<A, ?> ownerGetter,
                                                                        @NonNull final Function<A, Map<String, ?>> metaDataGetter) {
-        ParentPackageAnnotationScanner<A> parentPackageAnnotationScanner = new ParentPackageAnnotationScanner(annotationClass);
+        AnnotationScanner<A> parentPackageAnnotationScanner =
+                new CachedParentPackageAnnotationScanner<A>(
+                        new ParentPackageAnnotationScanner(annotationClass)
```
Comment (Jezorko) [inline src/main/java/com/glovoapp/ownership/OwnershipAnnotationDefinition.java:61]: [IDE] Please pay more attention to [your IDE warnings](https://www.jetbrains.com/help/idea/code-inspection.html). This code shows:

> Raw use of parameterized class 'ParentPackageAnnotationScanner' 

```suggestion
                new CachedParentPackageAnnotationScanner<>(
                        new ParentPackageAnnotationScanner<>(annotationClass)
```

ᴛʜɪs ᴀᴄᴛɪᴏɴ ᴡᴀs ᴘᴀʀᴛɪᴀʟʟʏ ᴀᴜᴛᴏᴍᴀᴛᴇᴅ ʙʏ [GɪᴛHᴜʙ's ʀᴇsᴘᴏɴsᴇ ᴛᴇᴍᴘʟᴀᴛᴇs](https://github.com/settings/replies)

File: src/main/java/com/glovoapp/ownership/scanning/AnnotationScanner.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.glovoapp.ownership.scanning;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AnnotatedElement;
+import java.util.Optional;
+
+public interface AnnotationScanner<A extends Annotation> {
+    Optional<A> scan(AnnotatedElement element);
```
Comment (Jezorko) [inline src/main/java/com/glovoapp/ownership/scanning/AnnotationScanner.java:8]: [nit] 
```suggestion
    Optional<A> scan(final AnnotatedElement element);
```

File: src/main/java/com/glovoapp/ownership/scanning/PackageScanningUtils.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package com.glovoapp.ownership.scanning;
+
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Method;
+import java.util.Optional;
+
+public class PackageScanningUtils {
```
Comment (Jezorko) [inline src/main/java/com/glovoapp/ownership/scanning/PackageScanningUtils.java:7]:
```suggestion
@NoArgsConstructor(access = AccessLevel.PRIVATE)
final class PackageScanningUtils {
```

File: src/main/java/com/glovoapp/ownership/scanning/ParentPackageAnnotationScanner.java [hunk 5]
```diff
@@ -1,23 +1,27 @@
...
-        String packageName = getPackageName(element).orElse("");
+        log.info("executing ParentPackageAnnotationScanner on element {}", element.toString());
+
+        String packageName = PackageScanningUtils.getPackageName(element).orElse("");
 
         while (packageName.length() > 0) {
-            packageName = getSuperPackageName(packageName);
+            packageName = PackageScanningUtils.getSuperPackageName(packageName);
 
             Class<?> packageInfo = null;
```
Comment (Jezorko) [inline src/main/java/com/glovoapp/ownership/scanning/ParentPackageAnnotationScanner.java:26]: [IDE] Please pay more attention to [your IDE warnings](https://www.jetbrains.com/help/idea/code-inspection.html). This code shows:

> Variable 'packageInfo' initializer 'null' is redundant 

```suggestion
            Class<?> packageInfo;
```

ᴛʜɪs ᴀᴄᴛɪᴏɴ ᴡᴀs ᴘᴀʀᴛɪᴀʟʟʏ ᴀᴜᴛᴏᴍᴀᴛᴇᴅ ʙʏ [GɪᴛHᴜʙ's ʀᴇsᴘᴏɴsᴇ ᴛᴇᴍᴘʟᴀᴛᴇs](https://github.com/settings/replies)

**Review by Jezorko — APPROVED** [review]

**Review by kjayachandra2000 — APPROVED** [review]

**Pull request merged**